src/ray/object_manager/test/pull_manager_test.cc

```diff
@@ -130,43 +143,71 @@ TEST_F(PullManagerTest, TestRestoreObjectFailed) {
   ASSERT_EQ(pull_manager_.NumActiveRequests(), 1);
 
   std::unordered_set<NodeID> client_ids;
-  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar");
+  pull_manager_.OnLocationChange(obj1, client_ids, "", NodeID::Nil());
 
   // client_ids is empty here, so there's nowhere to pull from.
   ASSERT_EQ(num_send_pull_request_calls_, 0);
-  ASSERT_EQ(num_restore_spilled_object_calls_, 1);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
 
-  restore_object_callback_(ray::Status::IOError(":("));
+  // Object is now spilled to a remote node, but the client_ids are still empty.
+  const NodeID remote_node_object_spilled = NodeID::FromRandom();
+  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar",
+                                 remote_node_object_spilled);
 
-  // client_ids is empty here, so there's nowhere to pull from.
   ASSERT_EQ(num_send_pull_request_calls_, 0);
   ASSERT_EQ(num_restore_spilled_object_calls_, 1);
 
-  client_ids.insert(NodeID::FromRandom());
-  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar");
-
-  // We always assume the restore succeeded so there's only 1 restore call still.
-  ASSERT_EQ(num_send_pull_request_calls_, 0);
-  ASSERT_EQ(num_restore_spilled_object_calls_, 1);
+  restore_object_callback_(ray::Status::IOError(":("));
 
+  // Now the restore request has failed, the remote object shouldn't have been properly
+  // restored.
   fake_time_ += 10.0;
-  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar");
+  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar",
+                                 remote_node_object_spilled);
 
   ASSERT_EQ(num_send_pull_request_calls_, 0);
   ASSERT_EQ(num_restore_spilled_object_calls_, 2);
 
-  restore_object_callback_(ray::Status::IOError(":("));
+  restore_object_callback_(ray::Status::OK());
+  // Now the remote restoration request succeeds, so we sholud be able to pull the object.
+  client_ids.insert(remote_node_object_spilled);
+  // Since it is the second retry, the interval gets doubled.
+  fake_time_ += 20.0;
+  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar",
+                                 remote_node_object_spilled);
 
-  // Since restore failed, we can fallback to pulling from another node immediately.
+  // Now that we've successfully sent a pull request, we need to wait for the retry period
+  // before sending another one.
   ASSERT_EQ(num_send_pull_request_calls_, 1);
   ASSERT_EQ(num_restore_spilled_object_calls_, 2);
+}
 
-  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar");
+TEST_F(PullManagerTest, TestLoadBalancingRestorationRequest) {
+  /* Make sure when the object copy is in other raylet, we pull object from there instead
```

Comment:
It would be nice to mimic the full restoration protocol. So first we receive just a spilled location, and check that we call restore. Then we receive the spilled location and a set of NodeIDs to pull from, so we should prioritize pulling.